### PR TITLE
Improve the "Argument must be a mapping" error message

### LINKED ISSUE #12070: Improve the "Argument must be a mapping" error message
**Feature**

This could also be classified as a bug report, but it's also a feature request. Consider code like this:

```py
from typing import Any, Optional

def test(**kwargs):
    ...

kwargs: Optional[Any]
test(**kwargs)
```

Now, this is obviously wrong on purpose, and in latest MyPy, results in this error message:

```
error: Argument after ** must be a mapping
```

The problem with this message is that, it doesn't say what exactly is wrong, unlike when you normally mismatch types where it says: `Argument 1 to "test" has incompatible type "Union[int, str]"; expected "bool"`, for example. In my case, I had a variable with the same name inferring `Optional[Any]` higher up in the code, and tracking down the root cause took me good 10 minutes,

**Pitch**

Even though `**kwargs` usage like this is uncommon, having a similarly improved message would help with debugging why this may be happening. Something like: `Argument after ** to "test" has incompatible type "Union[Any, None]", expected "Mapping[str, Any]"` should help a lot.

### PR DESCRIPTION
**Pull request by purna135**
### Description
Fixes #12070

Improved the error message "error: Argument after ** must be a mapping" when there is a mismatch types

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Could you add a test case where this makes a difference? I'd grep for existing test cases that use this message, and add another one that would be different without this PR.

**Comment by purna135** [comment]
Hi @JelleZijlstra, I'm completely new to this, so I'm a little perplexed right now.
I made the following changes to the check-kwargs.test file:

```
[case testInvalidTypeForKeywordVarArg]
from typing import Dict
from typing import Any, Optional
def f(**kwargs: 'A') -> None: pass
d = None # type: Dict[A, A]
f(**d)         # E: Keywords must be strings
f(**A())       # E: Argument after ** must be a mapping, not "A"
class A: pass
kwargs: Optional[Any]
f(**kwargs)    # E: Argument after ** must be a mapping, not "Union[Any, None]"
[builtins fixtures/dict.pyi]
```

But running the `pytest -n0 -k 'testInvalidTypeForKeywordVarArg'` gives the below output
```
E   AssertionError: Unexpected type checker output (C:\Users\Purna\Downloads\OpenSource\mypy\test-data\unit\check-kwargs.test, line 352)
------------------------------------------------------------------- Captured stderr call -------------------------------------------------------------------
Expected:
  main:5: error: Keywords must be strings
  main:6: error: Argument after ** must be a mapping, not "A"
  main:9: error: Argument after ** must be a mapping, not "Union[Any, None]" (diff)
Actual:
  main:5: error: Keywords must be strings
  main:6: error: Argument after ** must be a mapping, not "A"

================================================================= short test summary info ==================================================================
FAILED mypy/test/testcheck.py::TypeCheckSuite::check-kwargs.test::testInvalidTypeForKeywordVarArg
```

Please guide how to solve this, Thanks

**Comment by purna135** [comment]
Thanks @JelleZijlstra, this time I got some error : )
First of all there was a Incompatible types due to `d = None # type: Dict[A, A]`, so I removed this check temporarily to check my test.

Below is the error, what I got before this PR
**Test:**
```
[case testInvalidTypeForKeywordVarArg]
# flags: --strict-optional
from typing import Dict
from typing import Any, Optional
def f(**kwargs: 'A') -> None: pass
f(**A())       # E: Argument after ** must be a mapping, not "A"
class A: pass
kwargs: Optional[Any]
f(**kwargs)    # E: Argument after ** must be a mapping, not "Union[Any, None]"
[builtins fixtures/dict.pyi]
```

Output:
```
Expected:
  main:5: error: Argument after ** must be a mapping, not "A"
  main:8: error: Argument after ** must be a mapping, not "Union[Any, None]" (diff)
Actual:
  main:5: error: Argument after ** must be a mapping, not "A"
  main:8: error: Argument after ** must be a mapping (diff)

Alignment of first line difference:
  E: ...** must be a mapping, not "Union[Any, None]"
  A: ...** must be a mapping
```

After this PR, below is the output so I have to replace `Union[Any, None]` with `Optional[Any]`
```
Expected:
  main:5: error: Argument after ** must be a mapping, not "A"
  main:8: error: Argument after ** must be a mapping, not "Union[Any, None]" (diff)
Actual:
  main:5: error: Argument after ** must be a mapping, not "A"
  main:8: error: Argument after ** must be a mapping, not "Optional[Any]" (diff)

Alignment of first line difference:
  E: ...** must be a mapping, not "Union[Any, None]"
  A: ...** must be a mapping, not "Optional[Any]"
                                   ^
```

Shoul I replace the message with
```
self.fail(
      'Argument after ** has incompatible type {}, expected "Mapping[str, Any]"'.format(format_type(typ)),
      context, code=codes.ARG_TYPE)
```

**Comment by purna135** [comment]
and one more question, should I remove `# flags: --strict-optional` from test while commiting the changes ?

**Comment by purna135** [comment]
Hello, I'm waiting for your response; could you please direct me in the right direction?

**Comment by JelleZijlstra** [comment]
Hi, that's a lot of questions.

For `Union[Any, None]`, just change the test output to whatever variant makes it pass. We recently merged a PR that I think affects the output here.

I'm not sure why you are suggesting a message with `Mapping[str, Any]` in it. I feel like that would make the message harder to read for users.

And no, the `--strict-optional` should stay, because otherwise the tests will fail.

**Comment by purna135** [comment]
Thanks @JelleZijlstra 

`--strict-optional` causes the test to fail for below case with the message ` main:4: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[A, A]")`
 
```
def f(**kwargs: 'A') -> None: pass
d = None # type: Dict[A, A]
f(**d)         # E: Keywords must be strings
```

I am pushing the changes please have a look.

File: test-data/unit/check-kwargs.test [hunk 1]
```diff
@@ -350,12 +350,15 @@ class A: pass
 [builtins fixtures/dict.pyi]
 
 [case testInvalidTypeForKeywordVarArg]
-from typing import Dict
+# flags: --strict-optional
+from typing import Dict, Any, Optional
 def f(**kwargs: 'A') -> None: pass
 d = None # type: Dict[A, A]
```
Comment (JelleZijlstra) [inline test-data/unit/check-kwargs.test:356]:
```suggestion
d = {} # type: Dict[A, A]
```
Does this work?

**Comment by purna135** [comment]
Yes now it's passing all the tests, Thanks : )

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**